benchmark/tools/compare.py

```diff
@@ -0,0 +1,261 @@
+            benchmark_result = {}
+            warn_on_inconsistent_keys(
+                baseline[key], comparison[key], "{}.{}".format(context, key)
+            )
+            for benchmark_name in baseline[key].keys():
+                if isinstance(baseline[key][benchmark_name], dict):
+                    comparison_result = compare_benchmark(
+                        baseline[key][benchmark_name],
+                        comparison[key][benchmark_name],
+                        "{}.{}.{}".format(context, key, benchmark_name),
+                    )
+                    if len(comparison_result) > 0:
+                        benchmark_result[benchmark_name] = comparison_result
+            if len(benchmark_result) > 0:
+                result[key] = benchmark_result
+        else:
+            # everything that's not a dict should only depend on the key in the root level
+            if baseline[key] != comparison[key]:
+                print(
+                    "WARNING: Inconsistent value for {}: {} != {} in context {}".format(
+                        key, baseline[key], comparison[key], context
+                    ),
+                    file=sys.stderr,
+                )
+            result[key] = baseline[key]
+    return result
+
+
+def extract_benchmark_results(
+    input: dict, benchmarks: dict, case_key: tuple, context: str | None
+):
+    for key, value in input.items():
+        benchmark_name = key if context is None else "{}.{}".format(context, key)
+        if key in map(lambda x: x + suffix, comparison_keys):
+            benchmark_name = benchmark_name[: -len(suffix)]
+            if benchmark_name not in benchmarks.keys():
+                benchmarks[benchmark_name] = []
+            benchmarks[benchmark_name].append((case_key, value))
+        elif isinstance(value, dict):
+            extract_benchmark_results(value, benchmarks, case_key, benchmark_name)
+
+
+def is_outlier(value: float):
+    return math.fabs(math.log(value)) > math.log(1.0 + args.outlier_threshold / 100)
+
+
+baseline_json = parse_json_matrix(args.baseline)
+comparison_json = parse_json_matrix(args.comparison)
+warn_on_inconsistent_keys(baseline_json, comparison_json, "root")
+
+results = {}
+
+for key in baseline_json.keys():
+    results[key] = compare(baseline_json[key], comparison_json[key], key_to_str(key))
+
+outliers = {}
+if args.outliers:
+    benchmarks = {}
+    for key, value in results.items():
+        extract_benchmark_results(value, benchmarks, key, None)
+    for benchmark_name, benchmark_results in benchmarks.items():
+        outlier = sorted(
+            [
+                (case_key, value)
+                for case_key, value in benchmark_results
+                if is_outlier(value)
+            ],
+            key=lambda x: math.fabs(math.log(x[1])),
+            reverse=True,
+        )
+        outliers[benchmark_name] = outlier[: min(len(outlier), args.outlier_count)]
+
+if args.output == "json":
```

Comment:
why is the json output not handled by pandas?

Reply:
I have two output styles in mind: The JSON output can handle arbitrary nesting, the tabular style flattens everything. I want to preserve the nested structure for the JSON output, which is hard to represent in Pandas, since there is no `orient` matching this structure well.